jgrapht-core/src/main/java/org/jgrapht/generate/ComplementGraphGenerator.java

```diff
@@ -0,0 +1,116 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Generator which produces the <a href="http://mathworld.wolfram.com/GraphComplement.html">complement graph</a> of a given input graph.
+ * The complement $\overline{G}$ of a graph $G$ consists of the same vertices as $G$, but whose edge set consists
+ * of the edges not in $G$.
+ * <p>
+ * More formally, let $G = (V, E)$ be a graph and let $K$ consist of all 2-element subsets of $V$. Then $\overline{G} = (V, K \setminus E)$ is
+ * the complement of $G$, where $K \setminus E$ is the relative complement of $E$ in $K$. For directed graphs, the complement can be defined
+ * in the same way, as a directed graph on the same vertex set, using the set of all 2-element ordered pairs of $V$ in place
+ * of the set $K$ in the formula above.
+ * <p>
+ * The complement is not defined for multigraphs. If a multigraph is provided as input to this generator, it will be treated as if
+ * it is a simple graph.
+ *
+ * @author Joris Kinable
+ * @since Sept 2017
+ *
+ *
+ * @param <V> vertex type
+ * @param <E> edge type
+ */
+public class ComplementGraphGenerator<V,E> implements GraphGenerator<V,E,V>{
+
+    private final Graph<V,E> graph;
+    private final boolean generateSelfLoops;
+
+    /**
+     * Complement Graph Generator
+     * @param graph input graph
+     */
+    public ComplementGraphGenerator(Graph<V, E> graph) {
+        this(graph, false);
+    }
+
+    /**
+     * Complement Graph Generator.
+     * If the target graph allows self-loops the complement of $G$ may be defined by
+     * adding a self-loop to every vertex that does not have one in $G$. This behavior can be controlled
+     * using the boolean <code>generateSelfLoops</code>.
+     *
+     * @param graph input graph
+     * @param  generateSelfLoops indicator whether self loops should be generated.
+     *                           If false, no self-loops are generated, independent of whether the target graph supports self-loops.
+     */
+    public ComplementGraphGenerator(Graph<V, E> graph, boolean generateSelfLoops) {
+        this.graph = GraphTests.requireDirectedOrUndirected(graph);
+        this.generateSelfLoops=generateSelfLoops;
+    }
+
+    /**
+     * Generates the complement graph.
+     * @param target receives the generated edges and vertices; if this is non-empty on entry, the
+     *        result will be a disconnected graph since generated elements will not be connected to
+     *        existing elements
+     */
+    public void generateGraph(Graph<V, E> target){
+        this.generateGraph(target, null, null);
+    }
+
+    @Override
+    public void generateGraph(Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap) {
+        Graphs.addAllVertices(target, graph.vertexSet());
+
+        if (graph.getType().isDirected()) {
+            for (V u : graph.vertexSet())
+                for (V v : graph.vertexSet())
+                    if (u == v)
```

Comment:
Here better use `equals()`.

Reply:
Is equals() really necessary here? I iterate twice over the same set of vertices. The set is not modified in between. In this case, it seems impossible to have a case where u==v fails, but u.equals(v) is satisfied. The inner loop should only be skipped if u and v are physically the same object.

A simple "==" comparison is generally more efficient than an equals comparison.

Reply:
Right!